### ISSUE TITLE
All items in one array, bucket/column set with property

### ISSUE THREAD

**Issue by andrisi** [kristianmandrup/vue2-dragula#1]
As Vuejs is data based, it would be more apropriate to store all items in one array, and specify the bucket (column) they are in with a property. It's the same direction as vuelidate took over vee-validate. Each bucket in the template would have a v-if based on that field. Of course this has to be two-way synced, so if I set the .bucket property of an item, it would be moved to the apropriate bucket/column in the interface, and if I drag them around, the .bucket property shall be updated. Is this possible or planned? The current setup makes this compplicated.

Same as https://github.com/Astray-git/vue-dragula/issues/34

**Comment by kristianmandrup** [comment]
Hi @andrisi 

I'm not sure I understand your strategi from your description. I specifically designed the library to be much more flexible than vue-dragula, so that you can tweak it to fit your needs.
You can pass in various factory methods to create your own instances and you can subclass each of the classes or write your own that conforms to the same basic API.

Currently there is a `ModelManager` for the source and target respectively (see `DragHandler`), with methods to insert into target (`insertAt`), remove from source (`removeAt`). 
You can wrap whatever model you like, such as a single array and have the `DragHandler` return the same ModelManager (see `getModel`) for both source and target if that suits you better.

**Comment by andrisi** [comment]
I think my strategy is quite generic, so you could add it as a default way to handle data - and so keep dragula's dead-simple aproach. I see you are flexible enough for this to be implemented. Let me try to explain again. Let's say we want a drag&drop interface to group employees into teams. In Vue, we would have a single employee array like:

` [ {name: John", group: 0 }, { name: "Steve", group: 1 }, { name: "Carla", group: 1 } ]`

The Vue tamplate would have something like this for each group (perhaps build based on another array just containing group ids and their names:

`<ul v-for="employee in employees" v-if="user.group = 0"><li>{{ user.name }}</li></ul>`

And that's it. Dragula would just update the employee's group attribute as need, and the interface would update automatically if you'd change the underlying data.

**Comment by kristianmandrup** [comment]
Sure, I see your point. Then you just need a decorator, to decorate domain model data with grouping metadata. However I don't plan to implement it as a baked in strategy at this point. You are very welcome to fork the project and have a go at it. You would need to do a full rewrite of: `DragHandler`, `ModelManager` and parts of `Service`

**Comment by andrisi** [comment]
Grouping data is *not* metadata, it's just simple data, just as the name - mirroring how we represent grouping in the database. I feel that you're overcomplicating the solution. I'll look into it, but you're in a better position to do so as you know how all this is handled by dragula itself. I just pushing to do it the "Vue way" - based on data.

**Comment by kristianmandrup** [comment]
I'm intrigued, but tbh I'm not really an expert on the internals of dragula. That's where you come in!

I will try to help you out tweaking the current mechanics:

```js
// DragHandler

  drag (el, source) {
    this.log('drag', el, source)
    this.dragElm = el
    this.dragIndex = this.domIndexOf(el, source)
  }

  drop (dropEl, target, source) {
    // ...
    this.dropIndex = this.domIndexOf(dropEl, target)
    this.sourceModel = this.getModel(source)
    this.dropModel(dropEl, target, source)
  }

  getModel(location) {
    return this.modelManager.createFor({
      name: this.name,
      drake: this.drake,
      logging: this.logging,
      model: this.findModelForContainer(location, this.drake)
    })
  }
```

So you would return two different managers with same underlying array model, but divided into source/target manager via `group` item entry?
Then `dropIndex` and `dragIndex` would be in relation to the singular array used by both managers. 

All you would have to do is change the `insertAt` and `removeAt` manager methods to switch group of each item instead of physically moving them. Cool :)

In `DragulaService` constructor:

```js
    this.createModelManager = options.createModelManager || createModelManager
    this.modelManager = this.createModelManager(options)
```

in `ModelManager` constructor:

```js
    this.modelRef = opts.model || []
    this.model = this.createModel(this.modelRef)
```

So if you ensure the same `model` is passed in `opts`, it should be pretty easy to make it work I imagine. The `model` needs to be passed in when creating the service.

In `function updateDirective(container, binding, vnode, oldVnode) {` of `vue-dragula` directive function, we currently have:

```js
    if (!drake.models) {
      drake.models = []
    }
    // ...

      drake.models.push({
        model: newValue,
        container: container
      })
```

Which I think needs to be changed as well then?

Ideally the directive function should a function that creates a class instance with the internals divided into utility classes that can be customised/subclassed as needed.

**Comment by andrisi** [comment]
We would have to look into how dragula works, because I feel that moving html items is in general not compatible with the way Vue does things. You'd better not touch the html at all, and let Vue do it reacting to changes in the data. Perhaps only for visual feedback on the drag destination, with a predefined class, but not adding html code. I'll think about it. Maybe another drag&drop library is better suited for this.

**Comment by kristianmandrup** [comment]
Just made a major refactoring here:
https://github.com/kristianmandrup/vue2-dragula/tree/granular

But there is an error I have to track down:

```
Module parse failed: /Users/kristianmandrup/repos/vue-projs/vue2-dragula/dist/vue-dragula.js Unexpected token (1074:5)
You may need an appropriate loader to handle this file type.
SyntaxError: Unexpected token (1074:5)
```

```js
	var typeof = typeof Symbol === "function" && typeof Symbol.iterator === "symbol" ? function (obj) {
	  return typeof obj;
	} : function (obj) {
	  return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj;
	};
```

**Comment by andrisi** [comment]
I started playing with the HTML5 drag & drop API inside Vuew, will get back here when I understand it.

**Comment by kristianmandrup** [comment]
The new [granular](https://github.com/kristianmandrup/vue2-dragula/tree/granular) branch should make it MUCH easier to customise strategy, underlying model etc. as needed!

**Comment by kristianmandrup** [comment]
I've heard that the HTML5 Drag'n Drop API is very "unsatisfactory". 
Not sure if they have improved it since then...?!

**Comment by kristianmandrup** [comment]
The new branch looks to be working now and has more docs on the new more advanced extension/customisation options. Cheers!
When tested, this will be released as version 3.0

**Comment by andrisi** [comment]
I worked a few hours on my data-driven drag-and-drop idea. The HTML5 API is OK, and in the end, it might be doable in like 100 lines of code, just I need to get to know Vue's internals better to write a plugin. Maybe early next year.

**Comment by kristianmandrup** [comment]
You can look at my latest branch, then you should get an idea of how to write a plugin.

[directive creator](https://github.com/kristianmandrup/vue2-dragula/blob/granular/src/vue-dragula/directive/creator.js#L53)

**Comment by kristianmandrup** [comment]
Working on a new branch [draghandlers](https://github.com/kristianmandrup/vue2-dragula/tree/draghandlers) where the `DragHandler` has been split into multiple utility classes so it is easier to customise.

**Comment by kristianmandrup** [comment]
With the new architecture, this is how easy it will be

[grouped-models](https://github.com/kristianmandrup/vue2-dragula/blob/draghandlers/README.md#grouped-models)

Then pass the group property of the model items as an additional `groupProp` option (will be 'group' by default).

```js
import { ModelManager, GroupOps } from 'vue2-dragula'

function createModelManager(opts) {
  const GroupedModelManager = Object.assign(ModelManager, GroupOps)
  return new GroupedModelManager(opts)
}
this.$dragula.createService({
  name: 'grouped',
  groupProp: 'category', // default: 'group'
  // ...
  createModelManager
})
```

**Comment by kristianmandrup** [comment]
BTW: Thank you so much for the suggestion. Forced me to clean up and restructure the design to be much easier to customise and extend. Getting to the point where it would even be pretty easy to swap out Dragula with the pure HTML5 API or another drag'n drop framework if need be :)

**Comment by andrisi** [comment]
I'm happy about it, thanks. I feel that dragula could be diched, as a big portion of what it does is about html manipulation. I added a feature request to Vuejs that would make it even easier to build a clean implementation (https://github.com/vuejs/vue/issues/4588). There is one element that's hard without dragula: visual feedback of drop locations. I'd probably use some special directive for that, that would either allow things to be dropped on items or between them, and would only appear when needed. Or use some css class on them. We could chat aboput this later, perhaps when the Vuejs team responds to my feature request.

```html
<ul>
  <li v-drop="before">before first</li>
  <li v-drop="on" v-for="item in array" v-on:drop="drop($item, $event)">{{ item.name }}</li>
  <li v-drop="after">after items and last</li>
</ul>
```

**Comment by andrisi** [comment]
Vuejs already has the ability to tell the click/drag/etc event handlers the item that was used to build the even target. This below works like charm, so simplifying drag&drop shall be easy too.

```html
<li v-for="item in array" @click="democlick(item)">{{ item }}</li>
```

**Comment by kristianmandrup** [comment]
Thanks for all the input. This will be feasible and easy to do in the upcoming release.
You are most welcome to help out in the most recent branch. Almost there! Cheers!
